### PR TITLE
3x tweaks

### PR DESCRIPTION
**Pull request by manolama**

### CODE REVIEW

File: common/src/main/java/net/opentsdb/data/ZonedNanoTimeStamp.java [hunk 1]
```diff
@@ -76,53 +76,69 @@
...
+    epoch = Instant.from(timestamp).getEpochSecond();
+    nanos = Instant.from(timestamp).getNano();
   }
   
   public ZonedNanoTimeStamp(long epoch_millis, ZoneId zone) {
     if (zone == null) {
       throw new IllegalArgumentException("Zone cannot be null.");
     }
     timestamp = ZonedDateTime.ofInstant(Instant.ofEpochMilli(epoch_millis), zone);
+    epoch = Instant.from(timestamp).getEpochSecond();
```
Comment (smrutilal2) [inline common/src/main/java/net/opentsdb/data/ZonedNanoTimeStamp.java]: You can also do `millisec/1000` to minimize garbage. `Instant.from(timestamp)` would create a new instance of `Instant` every time. In fact twice for the current code. One for second and one more for nano.

Comment (manolama) [reply common/src/main/java/net/opentsdb/data/ZonedNanoTimeStamp.java]: Good catch. I still need to use the timesamp functions though as we need to convert from a zoned timestamp to the Unix epochs.

**Comment by arungupta** [comment]
what are the use cases where you need to use zoned timestamps @manolama?

**Comment by manolama** [comment]
Whenever we have to downsample based on a time zone, e.g. accounting for DST when downsampling over days or weeks.

File: common/src/main/java/net/opentsdb/pools/ArrayObjectPool.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  /**
+   * Attempts to claim an object from the pool immediately, allocating a new
+   * one if they're all in use. If the requested length is greater than the 
+   * configured pool array length, a new array is allocated and will be GC'd.
+   * 
+   * @param length The length of the array to return.
+   * @return A non-null pooled object.
+   */
+  public PooledObject claim(final int length);
```
Comment (sidhhu) [inline common/src/main/java/net/opentsdb/pools/ArrayObjectPool.java]: where is release happening?

Comment (manolama) [reply common/src/main/java/net/opentsdb/pools/ArrayObjectPool.java]: happens in the code that calls the `claim` method.

File: core/src/main/java/net/opentsdb/pools/BlockingQueueArrayObjectPoolFactory.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * Generates blocking queue object pools for arrays.
+ * 
+ * @since 3.0
+ */
+public class BlockingQueueArrayObjectPoolFactory extends BaseTSDBPlugin 
+  implements ArrayObjectPoolFactory {
+
+  @Override
+  public ObjectPool newPool(final ObjectPoolConfig config) {
+    return new BlockingQueueObjectPool(tsdb, config);
```
Comment (smrutilal2) [inline core/src/main/java/net/opentsdb/pools/BlockingQueueArrayObjectPoolFactory.java]: shouldn't this be `BlockingQueueArrayObjectPool ` instead

File: core/src/main/java/net/opentsdb/pools/BlockingQueueArrayObjectPool.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    }
+  }
+  
+  @Override
+  public PooledObject claim() {
+    PooledObject obj = pool.poll();
+    if (obj == null) {
+      tsdb.getStatsCollector().incrementCounter("objectpool.claim.miss", 
+          "pool", config.id());
+      return new LocalPooled(config.allocator().allocate(), false);
```
Comment (smrutilal2) [inline core/src/main/java/net/opentsdb/pools/BlockingQueueArrayObjectPool.java]: This seems to be creating an array that will be garbage collected.

Comment (manolama) [reply core/src/main/java/net/opentsdb/pools/BlockingQueueArrayObjectPool.java]: Right, if all of the objects in the pool are checked out then we fall back to allocation.

File: core/src/main/java/net/opentsdb/pools/BlockingQueueArrayObjectPoolFactory.java (hunk 3 above)
Comment (manolama) [reply core/src/main/java/net/opentsdb/pools/BlockingQueueArrayObjectPoolFactory.java]: yup good catch.

File: core/src/main/java/net/opentsdb/pools/BlockingQueueArrayObjectPool.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    @Override
+    public Object object() {
+      return obj;
+    }
+
+    @Override
+    public void release() {
+      if (was_pooled) {
+        if (!pool.offer(this)) {
+          LOG.warn("Failed to return a pooled object to the pool for " + config.id());
```
Comment (smrutilal2) [inline core/src/main/java/net/opentsdb/pools/BlockingQueueArrayObjectPool.java]: should decrement the count

Comment (manolama) [reply core/src/main/java/net/opentsdb/pools/BlockingQueueArrayObjectPool.java]: oh and I don't actually have a count here to decrement. The config does have it but that's just an initialization time value. I can add a "size" getter if we need it.

File: common/src/main/java/net/opentsdb/pools/ArrayObjectPoolFactory.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ */
+public interface ArrayObjectPoolFactory extends TSDBPlugin {
+
+  /**
+   * Instantiates a new pool with the given config.
+   * @param config A non-null config.
+   * @return A non-null pool or an exception if the config was missing required
+   * information.
+   */
+  public ObjectPool newPool(final ObjectPoolConfig config);
```
Comment (smrutilal2) [inline common/src/main/java/net/opentsdb/pools/ArrayObjectPoolFactory.java]: return type should be `ArrayObjectPool `

**Pull request merged**